From cb7043c7c44a10dc4a89813faee9357ae21ffac7 Mon Sep 17 00:00:00 2001
From: Scott Hanson <computergeek1507@users.noreply.github.com>
Date: Sat, 10 Jun 2023 00:27:06 -0400
Subject: [PATCH] Change DXF Export to use ext Style handle ID for 340 Dimstyle
 field. Dimstyle should use the Text Style handle ID not the name. (#1551)

http://docs.autodesk.com/ACD/2011/ENU/filesDXF/WS1a9193826455f5ff18cb41610ec0a2e719-7a53.htm
---
 libraries/libdxfrw/src/libdxfrw.cpp | 26 ++++++++++++++++++--------
 libraries/libdxfrw/src/libdxfrw.h   |  1 +
 2 files changed, 19 insertions(+), 8 deletions(-)

diff --git a/libraries/libdxfrw/src/libdxfrw.cpp b/libraries/libdxfrw/src/libdxfrw.cpp
index 8dba53f3b9..919aefe900 100644
--- a/src/libdxfrw.cpp
+++ b/src/libdxfrw.cpp
@@ -314,15 +314,17 @@ bool dxfRW::writeLayer(DRW_Layer *ent){
 
 bool dxfRW::writeTextstyle(DRW_Textstyle *ent){
     writer->writeString(0, "STYLE");
+    //stringstream cause crash in OS/X, bug#3597944
+    std::string name=ent->name;
+    transform(name.begin(), name.end(), name.begin(), toupper);
     if (!dimstyleStd) {
-        //stringstream cause crash in OS/X, bug#3597944
-        std::string name=ent->name;
-        transform(name.begin(), name.end(), name.begin(), toupper);
-        if (name == "STANDARD")
+        if (name == "STANDARD"){
             dimstyleStd = true;
+        }
     }
     if (version > DRW::AC1009) {
         writer->writeString(5, toHexStr(++entCount));
+        textStyleMap[name] = entCount;
     }
 
     if (version > DRW::AC1012) {
@@ -541,12 +543,20 @@ bool dxfRW::writeDimstyle(DRW_Dimstyle *ent){
     if ( version > DRW::AC1018 && ent->dimfxlon !=0 )
         writer->writeInt16(290, ent->dimfxlon);
     if (version > DRW::AC1009) {
-        writer->writeUtf8String(340, ent->dimtxsty);
+        std::string txstyname = ent->dimtxsty;
+        std::transform(txstyname.begin(), txstyname.end(), txstyname.begin(),::toupper);
+        if(textStyleMap.count(txstyname) > 0) {
+            int txstyHandle = (*(textStyleMap.find(txstyname))).second;
+            writer->writeUtf8String(340, toHexStr(txstyHandle));
+        }
     }
     if (version > DRW::AC1014) {
-        writer->writeUtf8String(341, ent->dimldrblk);
-        writer->writeInt16(371, ent->dimlwd);
-        writer->writeInt16(372, ent->dimlwe);
+        if(blockMap.count(ent->dimldrblk) > 0) {
+            int blkHandle = (*(blockMap.find(ent->dimldrblk))).second;
+            writer->writeUtf8String(341, toHexStr(blkHandle));
+            writer->writeInt16(371, ent->dimlwd);
+            writer->writeInt16(372, ent->dimlwe);
+        }
     }
     return true;
 }
diff --git a/libraries/libdxfrw/src/libdxfrw.h b/libraries/libdxfrw/src/libdxfrw.h
index 7deee2fb31..a79b2cfa03 100644
--- a/src/libdxfrw.h
+++ b/src/libdxfrw.h
@@ -150,6 +150,7 @@ class dxfRW {
     bool writingBlock;
     int elParts;  /*!< parts number when convert ellipse to polyline */
     std::unordered_map<std::string,int> blockMap;
+    std::unordered_map<std::string,int> textStyleMap;
     std::vector<DRW_ImageDef*> imageDef;  /*!< imageDef list */
 
     int currHandle;
